sound/soc/sof/sof-client.c

```diff
@@ -0,0 +1,47 @@
+// SPDX-License-Identifier: (GPL-2.0 OR BSD-3-Clause)
+//
+// Copyright(c) 2020 Intel Corporation. All rights reserved.
+//
+// Author: Ranjani Sridharan <ranjani.sridharan@linux.intel.com>
+//
+
+#include <linux/device.h>
+#include "sof-client.h"
+
+static LIST_HEAD(client_list);
+static DEFINE_MUTEX(sof_client_mutex);
+
+int sof_client_drv_register(const struct sof_client_drv *drv)
+{
+	/* add to list of SOF client drivers */
+	mutex_lock(&sof_client_mutex);
+	list_add(&drv->list, &client_list);
+	mutex_unlock(&sof_client_mutex);
+
+	/* register virtbus driver */
+	return virtbus_drv_register(drv->virtbus_drv);
+}
+EXPORT_SYMBOL_NS(sof_client_drv_register, SND_SOC_SOF_CLIENT);
+
+void sof_client_drv_unregister(const struct sof_client_drv *drv)
+{
+	/* unregister virtbus driver */
+	virtbus_drv_unregister(drv->virtbus_drv);
+
+	/* remove client drv from list */
+	mutex_lock(&sof_client_mutex);
+	list_del(&drv->list);
```

Comment:
what is the plan with this list? What purpose does it serve?

Reply:
hmm my original idea was to maintain a list of client devices that can be used to check say, which clients need the DSP to remain in D0 during suspend. But I suppose this list wont help. I should get rid of it. I can just go through the client devices in snd_sof_dev.

Reply:
@plbossart I've changed the list to be a list of client_devices. This is needed because I intend to keep the client members in snd_sof_dev like this (for example):

struct sof_client_dev ipc_test_client;
struct sof_client_dev sof_audio;
struct sof_client_dev sof_probes;

and so on.

This is mainly for that when the client device uses the APIs for IpC TX, I can use the container_of macro to acquire the snd_sof_dev from the virtbus_dev device that will be the caller. 

Do you think it is acceptable?

Reply:
Sorry @ranj063 I didn't get your point at all. When would you use this list again and what for?

Reply:
@plbossart during system suspend, when the SOF core makes decision to put the DSP in D0I3 or D3, it needs to know if any of the clients have the need for the DSP to be in D0I3. This is one example for when we'd use the list to iterate through all the clients.

Reply:
ah ok. 
But if it's a list used by the core, doesn't the core need to use the mutex to prevent any changes to the list while decisions are being made?
It feels like this is in the wrong location.

Reply:
@plbossart Yes, I am struggling to figure out whether it makes sense to have a global list or not. 
A global list made sense when this was done during the driver registration. But now that its not, I am not sure it is the right thing to do. Basically, this is my confusion: whether to keep a list of SOF client devices as a member of the SOF device or whether to keep the individual client devices are members of the SOF device. Let me think about this a bit more

Reply:
Having a list as part of the sof_dev allows you to have more dynamic/run-time creation of the devices. e.g. if you had a Kconfig enabled, or information from ACPI/Device Tree, or register values codec_mask=0x4, then the core creates a device and adds it to its list. Then on device creation you find a matching driver and problem solved.

Having all possible devices hard-coded in the structure doesn't seem very scalable to me.

Reply:
@plbossart I've updated the patches with a private list for the snd_sof_dev. I think the one thing I should also consider is making the client device registration platform specific.

Reply:
> making the client device registration platform specific.

You left me in the dust here @ranj063, what did you have in mind?

Reply:
@plbossart what I meant was if we were to register client devices based on ACPI/DT information for example, this would be platform specific. So instead of the core registering all the devices , I should consider adding a new platform-specific op that would register these devices.

For now, the only client is the dummy IPC client which is not platform-specific. But when it comes to adding say the HDMI client, it should be.

Reply:
do you mean to say we could have a platform device, with a driver probe upon detecting a specific ID, which would send use the _client_register().
If yes, how on earth would it get any information on sdev?

Reply:
oh no absolutely not. What I mean is this:
1. add a platform-specific op called sof_client_register()
2. In the op, detect the platform information such as the codec_mask = 0x4. If so, register the client_dev for HDMI in the platform-specific op instead of in the top-level SOF code.

So for example IMX or even BYT, we would not register this client but for APL/CML/TGL we would

Reply:
when we talk about ops, we generally mean that there would be a .byt_ops, and .apl_ops.

What you describe sounds more like a quirk handling, where we detect the platform in a single function and figure things out?

Reply:
@plbossart struct snd_sof_dsp_ops has an op called machine_register which determines the machine driver for each platform. What I am suggesting is a new op called maybe client_register() to determine which clients need to be registered for each platform.

Reply:
ok, now I get it. Yes that'd be useful to e.g. only create the probe with APL+ , etc. This should indeed not be a core decision.